### ISSUE TITLE
Web Translation API

### ISSUE THREAD

**Issue by domenic** [mozilla/standards-positions#1015]
## Request for Mozilla Position on an Emerging Web Specification

* Specification title: Web Translation API
* Specification or proposal URL (if available): https://github.com/WICG/proposals/issues/147
* Explainer URL (if available): https://github.com/explainers-by-googlers/translation-api
* Proposal author: @domenic

### Other information

This is an early-stage idea (hasn't even moved to WICG yet, although hopefully it will soon: https://github.com/WICG/proposals/issues/147).

However, we'd love early engagement from other implementers to make sure what we're designing fits well with their architecture.

**Comment by tomayac** [comment]
(TAG review in https://github.com/w3ctag/design-reviews/issues/948.)

**Comment by saschanaz** [comment]
FYI, we have published our own Translation API explainer: https://github.com/mozilla/explainers/blob/main/translation.md

**Comment by saschanaz** [comment]
We are negative on this proposal. While we plan to describe the reasons further in [our proposal](https://github.com/mozilla/explainers/blob/main/translation.md), briefly it's about:

1. The spec assumes a certain form of translation backend, exposing information about model availability, download progress, quotas, and usage prediction. We'd like to minimize the information exposure so that the implementation can be more flexible.
2. DOM tree translation is a powerful feature that is nontrivial to polyfill. We'd like to provide first-level support for the page translation use case.

**Comment by domenic** [comment]
Looking forward to hearing more!

> 1. The spec assumes a certain form of translation backend, exposing information about model availability, download progress, quotas, and usage prediction. We'd like to minimize the information exposure so that the implementation can be more flexible.

I don't think this is accurate. The specification is clear that all of these signals are not required, e.g. you could always return "available" and there's no download progress, or it explicitly allows infinite quotas with zero usage.

> 2\. DOM tree translation is a powerful feature that is nontrivial to polyfill. We'd like to provide first-level support for the page translation use case.

Do you have evidence of developer demand for DOM tree translation? We've not seen developers interested in this, since browsers have this functionality built in. Instead, they're interested in translating information that's specifically *not* in the DOM tree.

**Comment by collimarco** [comment]
I have made a similar proposal here:
https://github.com/WICG/proposals/issues/215

**Comment by saschanaz** [comment]
We still think many of the spec is assuming cloud backend based on AI which we are not interested in. We can be positive if the spec can be split and anything non-basic features can go to a separate extension spec, like what happened to [File System](https://fs.spec.whatwg.org/) and [File System Access](https://wicg.github.io/file-system-access/). The non-basic features include model availability, download progress, quotas, and usage prediction as described above.

That means for now the proposed position is not changing.

**Comment by xM8WVqaG** [comment]
While I agree that the current proposed spec is very cloud LLM/model centric, several of the non-basic features you just highlighted are still relevant to the offline NMT based Bergamot engine used in Firefox such as model availability and download progress. The quota for offline translation is unlimited and usage prediction is zero.

<img width="698" height="325" alt="Image" src="https://github.com/user-attachments/assets/34d524bd-27e5-4a84-8af6-a79586018a5a" />

Given Firefox already supports calling out to cloud-hosted solutions in the AI Chatbox sidebar, it's not too difficult to imagine enabling external translation providers as an option for languages that aren't supported in Bergamot, or for complex translation (such as mixed language content) that Bergamot doesn't handle.

The current workaround is using the `trialML` API to run MT models which, given their relatively small size to run in the browser, give poor results in comparison.

**Comment by tomayac** [comment]
> We still think many of the spec is assuming cloud backend based on AI which we are not interested in. 

I don't think that's true. While cloud-based translations would be possible and spec compliant, the spec mainly assumes language packs ("models") need to be downloaded for local translation. This approach is shared with Firefox even, as highlighted in this [Firefox 146 blog post](https://blog.mozilla.org/en/firefox/firefox-146/#:~:text=And%20once%20the%20language%20model%20is%20downloaded%2C%20translations%20even%20work%20offline.). As the download can take a moment, the spec has the concept of download progress. For local processing, the input quota is in fact `Infinity`. 

<img width="391" height="64" alt="Image" src="https://github.com/user-attachments/assets/71dd2b63-ba0f-4908-b48d-f6faa2a9ea30" />

**Comment by Christopher-Hayes** [comment]
A couple additional links:

MDN doc: https://developer.mozilla.org/en-US/docs/Web/API/Translator_and_Language_Detector_APIs
Chrome "Language Detection" doc: https://developer.chrome.com/docs/ai/language-detection
Chrome "Translation API" doc: https://developer.chrome.com/docs/ai/translator-api
CanIUse: https://caniuse.com/mdn-api_translator (Chrome officially added support in June 2025)

**Comment by sandstrom** [comment]
For other developers interested in this, the approach we'll be taking is to 'feature detect' this and fall back to calling a backend endpoint, where we'll proxy the request to an LLM for translation. 

With Chrome having a pretty high market share, most visitors will get the nice experience with an in-browser JS flow (quicker and better for privacy sometimes).

Others will fall back to backend-supported translations.

**Comment by zcorpan** [comment]
Negative per @saschanaz's comments. Discussion can still continue in the closed issue.

**Comment by xM8WVqaG** [comment]
I would be interested to know what Mozilla plans to do instead. I know there was a one-pager alternative spec published in April 2025 but to my understanding there hasn't been any implementation?

As an end user it's frustrating that from my perspective Mozilla's stance is just that they don't like the proposal and there is no obvious next steps. Two people have commented on this thread to suggest that perhaps the proposal has been misunderstood by the Mozilla team and neither of us were even replied to.

Although you've said the discussion can continue, now you've "closed" the issue that feels a bit pointless.

**Comment by sandstrom** [comment]
I agree, it's somewhat annoying.

You don't have to agree with every new web API proposal, of course. But it doesn't seem like anyone from Mozilla is even attempting to respond to comments here, including from the spec author @domenic.

LLMs is a large shift in technology, akin to mobile or the internet. It's great if the web spec can adopt and build out new APIs based on this new technology. While this is just a small subset of what LLMs can do, it's a pretty obvious one with a lot of utility. Browsers not being on top of this technology shift will just hurt the open-web (which Firefox is supposedly working to protect).

But instead of engaging in a constructive discussion about the spec, Firefox is just throwing out some arguments that aren't even congruent with themselves.

For example, above you are saying "We still think many of the spec is assuming cloud backend […] which we are not interested in." and then in [your own explainer](https://github.com/mozilla/explainers/blob/main/translation.md) you are proposing downloading models: "Add a DOM-based API […] to download the language model"

***

I don't have a horse in this race (not affiliated with neither *Google* nor *Firefox*), I'm just a simple developer interested in better translations in web APIs. But Firefox ought to at least be able to keep an intellectually honest discussion about the proposal and it's merits and downsides.

**Comment by renchap** [comment]
Chiming here with my Mastodon hat: we have a core feature of allowing users to translate Mastodon posts into their own language, which currently relies on backend API calls (that are forwarded to various translation services that can be configured by the server admin). This creates significant costs for server admins and we would love to be able to replace it with a local translation feature provided by the browser.

A DOM-based API does not look very attractive to us as our web frontend uses React, thus a vDOM, and any API that changes the DOM will not be usable for us. I am not a browser developer and I appreciate the concerns mentioned in this issue, but I believe that Mozilla should consider a JS API as well for complex apps.

**Comment by tomayac** [comment]
> This creates significant costs for server admins and we would love to be able to replace it with a local translation feature provided by the browser.

PRs based on the [Translation API](https://github.com/WICG/translation-api) that implements said local translation feature: https://github.com/mastodon/mastodon/pull/35369 (translation) and https://github.com/mastodon/mastodon/pull/35345 (language detection).

**Comment by jakearchibald** [comment]
@renchap I presume you get post body content from the server in some format. If a post contains rich content, like monospace, are you converting that to vdom?

**Comment by jakearchibald** [comment]
@renchap looking at https://mastodon.social/api/v1/statuses/115960589233121998, the content for a toot is already in HTML format. Are you converting that to a VDOM? If so, what's the benefit of that? Most implementations I see just use `dangerouslySetInnerHTML` and bypass VDOM.

**Comment by renchap** [comment]
ActivityPub statuses use HTML as a markup language, but it is not meant to be (always) displayed as it is on clients. You likely want to rename some classes (that might conflict with some classes from your UI), or change the rendered HTML based on the markup (for example the Mastodon web UI strips the last line if it's only hashtags, and displays it back with a specific component). 

In a planned redesign, we are also looking into replacing some links with rich components (for example a link to an account, like a mention, would be replaced by an interactive version of it)

**Comment by jakearchibald** [comment]
@renchap How will plain text translation work with rich content like the example post?

**Comment by renchap** [comment]
I am not sure I understand the question correctly, but our current translation API accepts HTML (content of the status) as input, and we get back translated HTML as the output. Then we run the resulting HTML through our React pipeline to transform it into vDOM, similar as we do for the HTML from the status.

I did not test it, but I believe that [@tomayac's PR linked above](https://github.com/mastodon/mastodon/pull/35369) shows that it works the same with the proposed JS-based translation API.

**Comment by jakearchibald** [comment]
I'm not seeing anything in the explainer that suggests that the API should/shouldn't work with content that's a mix of human language and some forms of markup. This would be a major source of interop and potentially security issues.

**Interop issues:** One engine supports translating text in HTML source. Another supports text in JSON files. Another supports neither of those. The spec says all of these are compliant.

**Security issues:** Due to one implementation, developers get used to piping HTML through this API, then parsing the result as HTML. Another implementation is not prepared for this, and can be tricked into producing HTML script tags that weren't in the original text.

Is that fair to say @tomayac?

**Comment by tomayac** [comment]
That's correct, the Translator API (by accident? see https://github.com/webmachinelearning/translation-api/issues/60) works quite well with HTML, which rendered translating rich content possible:

```js
const translator = await Translator.create({
  sourceLanguage: 'en',
  targetLanguage: 'es',
});

await translator.translate('<p>Hello, or should <a href="https://example.com">we</a> say <strong>good morning</strong>?</p>');
// '<p>Hola, o debería <a href="https://example.com">nosotros</a> decir <strong>Buenos días</strong>?</p>'
```

**Comment by jakearchibald** [comment]
That seems really risky.

If we need this to work with rich content, it seems much safer to work with already-parsed content, like a DOM tree. Maybe for Mastodon's use-case that means parsing, translating, serialising, and parsing as VDOM, which isn't ideal, but I can't see another safe way to do it.

Even if you create a special HTML-safe version of this API, you still hit the problem where it appears to "work quite well" in cases where the developer forgets to use the HTML-safe version, then things go wrong later.

**Comment by renchap** [comment]
For our own usage, as translation is not automatic and a one-time action (a user clicks on a link to get a translated version of a post), I believe that this would be all good if we could run `DOMParser` on the status HTML, then provide this DOM tree to a JS API, get another DOM tree as a result, and use then feed this to our React pipeline to get it rendered.

What would not work for us is only having an API that would in-place translate part of the DOM tree (as this would break React vDOM), but anything that allows us to do it in pure JS would work. And now that I think about it, in the above case we could have an ugly workaround where we render the status HTML into an invisible part of the page, get it translated, and read it back, but that's very inelegant.

Whatever solution that allows us to get us a locally-translated version of an HTML status will be a great improvement for our users (and cost reduction for admins). And our position here is that we would strongly prefer this API to be accepted by at least 2 browser engines before we implement it, as we strongly believe in open standards :)

**Comment by jakearchibald** [comment]
Yeah it would be something like:

```js
const tmp = document.createElement('div');
tmp.innerHTML = postContent;
await domTranslationAPIThingy(tmp, options);
const translatedHTML = tmp.innerHTML;
```

**Comment by jakearchibald** [comment]
> but that's very inelegant.

I get it. But a solution that results in a security issue due to inconsistent and unpredictable handling of HTML would be worse.

Besides, if an implementation decides the content is HTML, the safest thing would be to parse it, translate the text, and serialise the result. So, the same inelegance, just behind a curtain.

**Comment by renchap** [comment]
What I described as very inelegant was requiring for the content you want to translate to be added to the document's DOM (and being rendered). I believe that your solution does not require this, so it would be a valid alternative (from what I understand, I am not the front-end expert in our team).

We are well aware of the danger of displaying arbitrary HTML into a page, and I believe that we are parsing the statuses DOM somewhere to sanitise it due to past security issues around those areas.

**Comment by tomayac** [comment]
I think from an API design point of view, it should ideally match the ergonomics of a cloud API ([example](https://docs.cloud.google.com/translate/docs/translate-text#advanced_tx)), where you enable additional safety features by passing the translation request a MIME type (like `text/html`) and therefore can be sure the result is safe to insert on a page. Though I'd still run the result through the [HTML Sanitizer](https://developer.mozilla.org/en-US/docs/Web/API/HTML_Sanitizer_API) now that it's about to be part of the platform.   

```js
const translator = await Translator.create({
  sourceLanguage: 'en',
  targetLanguage: 'es',
  mimeType: 'text/html', // Or `inputFormat` or something
});

await translator.translate('<p>Hello, or should <a href="https://example.com">we</a> say <strong>good morning</strong> <img src="lol" onerror="alert(`yo`)>?</p>');

// '<p>Hola, o debería <a href="https://example.com">nosotros</a> decir <strong>Buenos días</strong>?</p>'
```

**Comment by jakearchibald** [comment]
Yeah, but that has the following problem:

> Even if you create a special HTML-safe version of this API, you still hit the problem where it appears to "work quite well" in cases where the developer forgets to use the HTML-safe version, then things go wrong later.

**Comment by tomayac** [comment]
Could this maybe be mitigated by the API converting HTML to plain text (`<` to `&lt;` etc.) if the input is detected to be HTML (i.e., if it contains even a single tag), but the developer forgot to specify the HTML MIME type? This way `innerHTML`'ing (and friends) such a string would very quickly visibly fail.

**Comment by xM8WVqaG** [comment]
> I think from an API design point of view, it should ideally match the ergonomics of a cloud API ([example](https://docs.cloud.google.com/translate/docs/translate-text#advanced_tx)), where you enable additional safety features by passing the translation request a MIME type (like text/html) and therefore can be sure the result is safe to insert on a page.

I am not sure if you intentionally meant for this suggestion to be extensible, but a field like `mimeType` would enable some additional functionality. For example, passing in `image/svg+xml` or even `image/png` which are both formats that can currently be translated today by some translation tools.

Given the API is just called _Web Translation API_ and not _Text Translation API_, I don't think it's unreasonable to expect to be able to provide input formats other than `text/plain` in the future. Just like today, where different browsers support different images, videos, audio, containers, etc - and rendering isn't 100% consistent in all browsers - it would be unreasonable to expect something very subjective like a translation API to be 100% consistent between browsers.

**Comment by jakearchibald** [comment]
> it would be unreasonable to expect something very subjective like a translation API to be 100% consistent between browsers.

Yes but security isn't as subjective, and correctly handling HTML is very important for security.

Do we have other text-to-text transforming APIs on the platform that are HTML safe? Things like `str.toUpperCase()` are not.

> Given the API is just called _Web Translation API_ and not _Text Translation API_, I don't think it's unreasonable to expect to be able to provide input formats other than `text/plain` in the future.

Right, but `text/plain` and `text/html` are both text. That's the problem here.

**Comment by jakearchibald** [comment]
@tomayac 

> Could this maybe be mitigated by the API converting HTML to plain text (`<` to `&lt;` etc.) if the input is detected to be HTML

Only if this part were specified so it could be interoperable and reliable.

Alternatively, a DOM tree doesn't have format ambiguity, clearly differentiates human vs computer content, and doesn't carry the same risks.

**Comment by tomayac** [comment]
> > Could this maybe be mitigated by the API converting HTML to plain text (`<` to `&lt;` etc.) if the input is detected to be HTML
> 
> Only if this part were specified so it could be interoperable and reliable.

This seems feasible… I will add a note to https://github.com/webmachinelearning/translation-api/issues/60. 

> Alternatively, a DOM tree doesn't have format ambiguity, clearly differentiates human vs computer content, and doesn't carry the same risks.

Right. Do you envision an API shape, though, that always requires a DOM tree and that wouldn't have a simple text in, text out mode (as the current API)?

**Comment by martinthomson** [comment]
Relevant to this, which might strengthen Jake's point, the sanitizer API is now widely available.  https://frederikbraun.de/why-sethtml.html explains some of why `el.innerHTML = someText` is better spelled as `el.setHTML(someText)`.

**Comment by jakearchibald** [comment]
@tomayac 

> Do you envision an API shape, though, that always requires a DOM tree and that wouldn't have a simple text in, text out mode (as the current API)?

I'm not sure. If the text API encourages folks to assume it can handle HTML (as you did in the mastodon PR), when that guarantee doesn't actually exist, then it's unsafe.

**Comment by jakearchibald** [comment]
Fwiw, I just tried this in Chrome Canary:

```js
const translator = await Translator.create({
  sourceLanguage: 'en',
  targetLanguage: 'fr',
  monitor(m) {
    m.addEventListener('downloadprogress', (e) => {
      console.log(`Downloaded ${e.loaded * 100}%`);
    });
  },
});

await translator.translate('<div>I want to know the full <details>story</details></div>');
```

And it gives me `<div>Je veux connaître les <détails>histoires</details></div>`. So yeah, any handling of HTML in the current API is coincidental and not safe.

**Comment by jakearchibald** [comment]
```js
const translator = await Translator.create({
  sourceLanguage: 'fr',
  targetLanguage: 'en',
  monitor(m) {
    m.addEventListener('downloadprogress', (e) => {
      console.log(`Downloaded ${e.loaded * 100}%`);
    });
  },
});

await translator.translate('<scripte>console.log("hello")</scripte>')
```

Becomes `<script>console.log("hello")</script>` (the tags have become valid script tags).

**Comment by jakearchibald** [comment]
In summary: It seems there are use-cases for this API to handle forms of 'rich text', such as DOM/HTML. Without that use-case handled, it seems risky to ship the text-only API, as it encourages developers to just throw text at it, which becomes an attack vector.

**Comment by sandstrom** [comment]
There are many ways you can write unsafe code in the browser. We're currently implementing the translator API, and would never pipe it directly into one of the unsafe sinks. 

I think XSS safe APIs is in a pretty good state 2026. We have CSP, Trusted Types, the Sanitizer API, etc.

If it was up to me, just clearly labeling this risk in MDN docs etc would be enough.

**Comment by jakearchibald** [comment]
Right, but if many users want a way to support 'rich text', but we don't provide a way to handle that (despite it being something the browser already does), then it encourages folks to do the unsafe thing to get what they want, because sometimes it looks like it works. Like, we've seen it happen in this thread with the Mastodon PR.

> just clearly labeling this risk in MDN docs etc would be enough.

If there's a text-only API, we'd absolutely do this. But it seems bad if the advice is "if you have rich content, don't use this! There's no alternative for you, sorry".

**Comment by jakearchibald** [comment]
Here's a fun one:

```js
const translator = await Translator.create({
  sourceLanguage: 'zh',
  targetLanguage: 'en',
  monitor(m) {
    m.addEventListener('downloadprogress', (e) => {
      console.log(`Downloaded ${e.loaded * 100}%`);
    });
  },
});

await translator.translate(`Hello〈 脚本>console.log('hello')〈 /脚本>`);
// hello<script>console.log('hello')</script>
```

This is interesting because the source correctly parses as a single text node, but the translator returns a script tag. So, filtering HTML before it hits the translator isn't an option.

**Comment by collimarco** [comment]
This proposal is very complex and for advanced use cases.

For basic websites and pages, can we simply expose a `window.showTranslationPrompt();` as described here?

https://github.com/WICG/proposals/issues/215

Almost all browsers already have that feature available. 

The problem is that you have to go to `Menu (⋮) -> Translate` to trigger that prompt. Many non technical users don't see that.

The only addition would be to **allow websites to trigger that translation prompt after a click on a button**, instead of having the users search in the browser menu.

It would be extremely simple to add and very useful for users and websites.
